Runtime/Scripts/EdgeEventsManager.cs

```diff
@@ -609,17 +603,100 @@ IEnumerator UpdateLocation()
       else
       {
         yield return StartCoroutine(LocationService.EnsureLocation());
+        location = LocationService.RetrieveLocation();//throws location exception if location is not retrieved
+        yield break;
+      }
+    }
+
+    void ProcessLatency(Statistics stats)
+    {
+      Logger.Log("Processing Latency Received from Server");
+      if (stats.Avg > config.latencyThresholdTriggerMs)
+      {
+        Logger.Log("Current Latency (" + stats.Avg + ") > LatencyThreshold (" + config.latencyThresholdTriggerMs + ")");
+        Logger.Log("Performing FindCloudlet PerformanceMode");
+        PlatformIntegration pIntegration = new PlatformIntegration();
+        MatchingEngine matchingEngine = new MatchingEngine(pIntegration.CarrierInfo, pIntegration.NetInterface, pIntegration.UniqueID, pIntegration.DeviceInfo);
+        FCPerformanceThread tws = new FCPerformanceThread
+          (matchingEngine: matchingEngine, sessionCookie: integration.matchingEngine.sessionCookie, location: location, hostOverride: hostOverride, portOverride: portOverride, callbackDelegate: FCCallback);
         try
         {
-          location = LocationService.RetrieveLocation();//throws location exception if location is not retrieved
+          FCPerformanceThread = new Thread(new ThreadStart(tws.RunFCPerformance));
+          FCPerformanceThread.Name = "FCPerformanceThread";
+          FCPerformanceThread.Start();
+          FCPerformanceThread.IsBackground = true;
+          FCPerformanceThread.Priority = System.Threading.ThreadPriority.Lowest;
```

Comment:
should probably call Start() after all configurations set

Reply:
Also, it is confusing that FCPerformanceThread is a variable name for a Thread and the name of one of your FCPerformanceThread class. 

I think what might make sense is to stick the actual Thread (new Thread(...)) variable into your FCPerformanceThread class and start the thread from inside a class method. Something like this:

```
class FCPerformanceThread
{
    Thread thread;
    string hostOverride;
    uint portOverride;
    string sessionCookie;
    MatchingEngine matchingEngine;
    EdgeEventsManager.FCPerformanceCallback callback;
    Loc location;
.
.
.
    internal async void RunFCPerformance() 
    {
          thread = new Thread(new ThreadStart(RunFCPerformanceHelper));
          thread.Name = "FCPerformanceThread";
          thread.IsBackground = true;
          thread.Priority = System.Threading.ThreadPriority.Lowest;
          thread.Start();
          thread.Join();
    }
    internal async void RunFCPerformanceHelper { (code that you have in RunFCPerformance right now) }
}
```

Reply:
I get it that the naming is confusing, but I need to spawn the thread from the UI thread (Update Monobehviour) in this case.

Reply:
Won't this still be run on the UI thread if you call RunFCPerformance from ProcessLatency?

Reply:
ah I guess you can make RunFCPerformance not an async method, then that should work fine:

```
internal void RunFCPerformance() 
    {
          thread = new Thread(new ThreadStart(RunFCPerformanceHelper));
          thread.Name = "FCPerformanceThread";
          thread.IsBackground = true;
          thread.Priority = System.Threading.ThreadPriority.Lowest;
          thread.Start();
          thread.Join();
    }
```

Reply:
In ProcessLatency I am spawning a thread then
I am calling RunFCPerformance from inside the Thread as a [ParameterizedThreadStart Delegate](https://docs.microsoft.com/en-us/dotnet/api/system.threading.parameterizedthreadstart?view=net-5.0)

Reply:
I did an update now, let me know if it does make the impl. more clear

Reply:
hmm, not a huge fan of RunFCPerformanceClass name. 

Calling fcthread.RunFCPerformance() should do the exact same thing as what you are doing now. It will just keep all of the Thread variables/code all inside inside one class.

Also you probably don't need a Thread class variable at all, it can just be a local variable.

```
void ProcessLatency(Statistics stats)
{
      Logger.Log("Processing Latency Received from Server");
      if (stats.Avg > config.latencyThresholdTriggerMs)
      {
        Logger.Log("Current Latency (" + stats.Avg + ") > LatencyThreshold (" + config.latencyThresholdTriggerMs + ")");
        Logger.Log("Performing FindCloudlet PerformanceMode");
        PlatformIntegration pIntegration = new PlatformIntegration();
        MatchingEngine matchingEngine = new MatchingEngine(pIntegration.CarrierInfo, pIntegration.NetInterface, pIntegration.UniqueID, pIntegration.DeviceInfo);
        FCPerformanceThreadClass fcThread = new FCPerformanceThreadClass
          (matchingEngine: matchingEngine, sessionCookie: integration.matchingEngine.sessionCookie, location: location, hostOverride: hostOverride, portOverride: portOverride, callbackDelegate: FCCallback);
        **fcThread.RunFCPerformance();**
.
.
.
}


class FCPerformanceThread
{
    string hostOverride;
    uint portOverride;
    string sessionCookie;
    MatchingEngine matchingEngine;
    EdgeEventsManager.FCPerformanceCallback callback;
    Loc location;
.
.
.
    internal void RunFCPerformance() 
    {
          Thread thread= new Thread(new ThreadStart(**RunFCPerformanceHelper**));
          try
          {
              thread.Name = "FCPerformanceThread";
              thread.IsBackground = true;
              thread.Priority = System.Threading.ThreadPriority.Lowest;
              thread.Start();
              Debug.Log("Waiting for FindCloudlet Performance Mode");
              thread.Join();
          }
          catch (ThreadInterruptedException tie)
          {
              thread.Abort();
              Logger.Log("FCPerformance Thread interrupted " + tie.Message + ", InnerExcetion: " + tie.InnerException + ", Stack: " + tie.StackTrace);
              PropagateError(FindCloudletEventTrigger.LatencyTooHigh, "Error In FindCloudlet Perfromance Mode, Thread Interrupted");
              return;
        }
    }

    internal async void RunFCPerformanceHelper { 
      FindCloudletReply fcReply = null;
      matchingEngine.EnableEdgeEvents = false;
      matchingEngine.sessionCookie = sessionCookie;
      FindCloudletRequest fcReq = matchingEngine.CreateFindCloudletRequest(location, "");
      Debug.Log("FindCloudletPerformanceMode Request: " + fcReq.ToString());
      try
      {
        fcReply = await matchingEngine.FindCloudletPerformanceMode(hostOverride, portOverride, fcReq, 5);
      }
      catch (Exception fce)
      {
        Debug.Log("FindCloudletException: " + fce.Message + "Inner Exception: " + fce.InnerException + ", Stack: " + fce.StackTrace);
        callback(fcReply);
        Thread.CurrentThread.Abort();
      }
      Debug.Log("FindCloudletPerformance done :" + fcReply.Status);
      if (callback != null)
      {
        callback(fcReply);
      }
    }
}

```

Does this make sense or am I missing something?

Reply:
Lets have a call maybe